tests/test_beta_program_initiator.py

```diff
@@ -0,0 +1,403 @@
+    assert beta_program_initiator.requests(0)[RITUAL_ID_SLOT] == 1
+    assert beta_program_initiator.requests(1)[RITUAL_ID_SLOT] == 0
+    balance_after = token.balanceOf(beta_program_initiator.address)
+    assert balance_before - balance_after == ritual_cost_1
+    assert token.balanceOf(coordinator.address) == ritual_cost_2 + ritual_cost_1
+
+    assert coordinator.getRitualsLength() == 2
+    assert coordinator.getProviders(1) == nodes_1
+    ritual = coordinator.rituals(1)
+    assert ritual[0] == beta_program_initiator.address
+    assert ritual[1] == authority_1
+    assert ritual[2] == duration_1
+    assert ritual[3] == access_controller_1
+    assert ritual[4] == 1
+    assert ritual[5] == ritual_cost_1
+
+    events = beta_program_initiator.RequestExecuted.from_receipt(tx)
+    assert events == [beta_program_initiator.RequestExecuted(0, 1)]
+
+
+def test_refund(accounts, beta_program_initiator, token, coordinator, executor):
+    (
+        initiator_1,
+        initiator_2,
+        authority,
+        node_1,
+        node_2,
+        access_controller,
+        *everyone_else,
+    ) = accounts[2:]
+
+    nodes = [node_1, node_2]
+    duration_1 = DAY_IN_SECONDS
+    ritual_cost_1 = coordinator.getRitualInitiationCost(nodes, duration_1)
+    duration_2 = 2 * duration_1
+    ritual_cost_2 = coordinator.getRitualInitiationCost(nodes, duration_2)
+
+    token.mint(initiator_1, 10 * ritual_cost_1, sender=initiator_1)
+    token.approve(beta_program_initiator.address, 10 * ritual_cost_1, sender=initiator_1)
+    token.mint(initiator_2, 10 * ritual_cost_2, sender=initiator_2)
+    token.approve(beta_program_initiator.address, 10 * ritual_cost_2, sender=initiator_2)
+
+    # Can't refund non-existent request
+    with ape.reverts("Non-existent request"):
+        beta_program_initiator.refundFailedRequest(0, sender=executor)
+
+    # Register three requests
+    beta_program_initiator.registerInitiationRequest(
+        nodes, authority, duration_1, access_controller, sender=initiator_1
+    )
+    beta_program_initiator.registerInitiationRequest(
+        nodes, authority, duration_2, access_controller, sender=initiator_2
+    )
+    beta_program_initiator.registerInitiationRequest(
+        nodes, authority, duration_2, access_controller, sender=initiator_1
+    )
+
+    # Can't refund not executed request
+    with ape.reverts("Request is not executed"):
+        beta_program_initiator.refundFailedRequest(0, sender=executor)
+    beta_program_initiator.cancelInitiationRequest(2, sender=initiator_1)
+    with ape.reverts("Request is not executed"):
+        beta_program_initiator.refundFailedRequest(2, sender=initiator_2)
+
+    # Can't refund not failed request
+    beta_program_initiator.executeInitiationRequest(1, sender=executor)
+    beta_program_initiator.executeInitiationRequest(0, sender=executor)
+    request_0_ritual_id = 1
+    request_1_ritual_id = 0
+    for state in [0, 1, 2, 5, 6]:
```

Comment:
Nice - although can these state values use the named state values in the Coordinator contract for RitualStatus?

Reply:
yep, I was a bit lazy, will do

Reply:
✔️